File: NC/buy-and-sell-crypto.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:

        maxAtRightFromPos = [0 for i in range(len(prices))]

        currMax = -1

        for i in range(len(prices)-2, -1, -1):
            maxAtRightFromPos[i] = max(prices[i+1], currMax)
            currMax = max(currMax, maxAtRightFromPos[i])

        ans = 0

        for i in range(len(prices)-1):
            ans = max(ans, maxAtRightFromPos[i]-prices[i])

        return ans


class SolutionWithBetterSpaceComplexity:
    def maxProfit(self, prices: List[int]) -> int:

        l, r = 0, 1
        ans = 0

        while r < len(prices):
            if prices[l] < prices[r]:
                ans = max(ans, prices[r] - prices[l])
            else:
                l = r
            r += 1

        return ans

File: NC/buy-and-sell-crypto.cpp
class Solution
{
public:
    int maxProfit(vector<int> &prices)
    {
        int ans = 0;

        for (int i = 0, j = 1; j < prices.size();)
        {
            int diff = prices[j] - prices[i];
            if (diff < 0)
            {
                i = j;
                j = j + 1;
            }
            else if (diff > ans)
            {
                ans = diff;
                j++;
            }
            else
            {
                j++;
            }
        }
        return ans;
    }
};
